staging/src/k8s.io/client-go/tools/clientcmd/client_config.go

```diff
@@ -577,20 +577,17 @@ func (config *inClusterClientConfig) ClientConfig() (*restclient.Config, error)
 }
 
 func (config *inClusterClientConfig) Namespace() (string, bool, error) {
-	// This way assumes you've set the POD_NAMESPACE environment variable using the downward API.
-	// This check has to be done first for backwards compatibility with the way InClusterConfig was originally set up
-	if ns := os.Getenv("POD_NAMESPACE"); ns != "" {
-		return ns, false, nil
+	//get the namespace of current pod, if available
+	namespace, err := restclient.InClusterNamespace()
+	if err != nil {
+		klog.Error("Fail to get in-cluster namespace: ", err)
+		return "default", false, nil
 	}
-
-	// Fall back to the namespace associated with the service account token, if available
-	if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
-		if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
-			return ns, false, nil
-		}
+	// for compatibility, return default if the namespace is empty
+	if len(namespace) == 0 {
```

Comment:
is there a reason not to do this inside InClusterNamespace if all callers would have to do the same thing?